File: app/src/main/java/io/github/qwqgong/androidcyaml/runtime/RuntimeStateBus.java
package io.github.qwqgong.androidcyaml;

import android.os.Handler;
import android.os.Looper;

import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

final class RuntimeStateBus {
    interface Listener {
        void onRuntimeStateChanged(RuntimeSnapshot snapshot);
    }

    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final Set<Listener> listeners = new CopyOnWriteArraySet<>();
    private volatile RuntimeSnapshot snapshot = RuntimeSnapshot.stopped();

    void addListener(Listener listener) {
        listeners.add(listener);
        RuntimeSnapshot current = snapshot;
        mainHandler.post(() -> listener.onRuntimeStateChanged(current));
    }

    void removeListener(Listener listener) {
        listeners.remove(listener);
    }

    RuntimeSnapshot snapshot() {
        return snapshot;
    }

    void publish(RuntimeSnapshot next) {
        snapshot = next;
        mainHandler.post(() -> {
            for (Listener listener : listeners) {
                listener.onRuntimeStateChanged(next);
            }
        });
    }
}

File: app/src/main/java/io/github/qwqgong/androidcyaml/runtime/RuntimeStateBus.kt
package io.github.qwqgong.androidcyaml

import android.os.Handler
import android.os.Looper
import java.util.concurrent.CopyOnWriteArraySet

class RuntimeStateBus {
    fun interface Listener {
        fun onRuntimeStateChanged(snapshot: RuntimeSnapshot)
    }

    private val mainHandler = Handler(Looper.getMainLooper())
    private val listeners = CopyOnWriteArraySet<Listener>()

    @Volatile
    private var currentSnapshot: RuntimeSnapshot = RuntimeSnapshot.stopped()

    fun addListener(listener: Listener) {
        listeners.add(listener)
        val current = currentSnapshot
        mainHandler.post { listener.onRuntimeStateChanged(current) }
    }

    fun removeListener(listener: Listener) {
        listeners.remove(listener)
    }

    fun snapshot(): RuntimeSnapshot = currentSnapshot

    fun publish(next: RuntimeSnapshot) {
        currentSnapshot = next
        mainHandler.post {
            for (listener in listeners) {
                listener.onRuntimeStateChanged(next)
            }
        }
    }
}
